### PR TITLE
Mono-repoize Calyx LSP + Tree-sitter

### PR DESCRIPTION
**Pull request by sgpthomas**
I've decided that this is at a point where it makes sense to merge the LSP stuff into the main Calyx mono-repo. It makes dealing with using the Calyx lib for diagnostics easier.

### CODE REVIEW

File: calyx-lsp/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+serde.workspace = true
+serde_json.workspace = true
+tower-lsp = "0.20"
+tracing = "0.1.40"
+tracing-subscriber = "0.3.18"
+tree-sitter = "0.20"
+
+[dependencies.tokio]
+version = "1"
+features = ["io-util", "io-std", "macros", "rt-multi-thread", "net"]
```
Comment (rachitnigam) [inline calyx-lsp/Cargo.toml:35]: Multi-threading! In 2024!!

**Comment by rachitnigam** [comment]
Looks awesome! I will ask you to document things to your discretion so we can fix breakages when they happen. It might be good to provide a high-level overview in the `docs/` as well for people trying to build upon it. This will also make it easier for @EclecticGriffin and co to integrate the `cider-dap` stuff they've been working on for a nice, out-of-the-box VS Code extension

**Comment by sgpthomas** [comment]
I think that this is at a point where it's ready to review / merge into main. I've setup package publishing for the VSCode extension so that it's installable from the VSCode store. In the future it would be nice if it good automatically install the `calyx-lsp`. How does the Cargo package publishing stuff work? Is there anything special that I need to do to get `calyx-lsp` published on `crates.io`?

**Comment by sampsyo** [comment]
AMAZING. I looked things over briefly and I think everything's good to go!

I don't know much about the crates.io publishing; I imagine that @rachitnigam just manually types `cargo publish`? It would of course be cool to someday set up an action to do this automatically when stuff is tagged (or maybe I misremember and we already have that). But that can of course happen separately!

**Comment by rachitnigam** [comment]
Indeed! I just manually push these crates to cargo but it is high time for us to automate this process because it's blocking us from producing new releases quickly.

**Comment by rachitnigam** [comment]
I'm enabling auto-merge to we can prepare this for the release.

**Comment by rachitnigam** [comment]
Also, should we archive the `calyx-lsp` repo after merging this?

**Pull request merged**